contributors/design-proposals/pod-priority-api.md

```diff
@@ -0,0 +1,203 @@
+* Best effort / opportunistic
+
+### Resolving priority names 
+
+User requests sent to Kubernetes may have "PriorityName"  in their PodSpec.
+Admission controller resolves a PriorityName to its corresponding number and
+populates the "Priority" field of the pod status. The rest of Kubernetes
+components look at the "Priority" field of pod status and work with the integer
+value. In other words, "PriorityName" will be ignored by the rest of the
+system.
+
+### Modifying priority names 
+
+Admission controller can optionally receive a ConfigMap from the API server
+that defines a list of new priority names and their numbers or changes the
+integral value of existing ones. New configurations cannot introduce priority
+names mapped to a value higher than "system" and cannot change the value of
+"system" priority. The reason for this restriction is that kubernetes critical
+system pods will have "system" priority and no pod should be able to preempt
+them.
+
+Alternatively, we may ship Kubernetes with only the highest priority level
+hard-coded. The rest of the priority names are defined in a ConfigMap which
+could be shipped with Kubernetes.
+
+### Ordering of priorities 
+
+As mentioned earlier, a PriorityName is resolved by the admission controller to
+its integral value and Kubernetes components use the integral value. The higher
+the value, the higher the priority.
+
+### Dynamic change of priority names 
+
+Kubernetes supports change of priority names at runtime, however the changes
+won't change the priority of existing pods. As mentioned earlier, admission
+controller resolves priority names and populates the "Priority" field of pod
+specs. A new ConfigMap may be fed to the admission controller to define new
+priority names or change the integer value of existing names (other than
+"system"). These new changes affects only the pods created after the changes
+and existing pods will keep their existing Priority.
+
+#### Drawbacks of changing priority names 
+
+While Kubernetes allows changing priority names, it should be done only when
+absolutely needed. Changing priority names has the following disadvantages:
+
+
+
+*   May remove config portability: pod specs written for one cluster are no
+    longer guaranteed to work on a different cluster if the same priority names
+    do not exist on the new cluster. 
+*   If quota is specified for existing priorities (at the time of this writing,
+    we don't have this feature in Kubernetes), adding or changing priorities
+    will require reconfiguration of quota allocations.
+
+### Priority and QoS classes 
+
+Kubernetes has [three QoS
+classes](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-qos.md#qos-classes)
+which are derived from request and limit of pods. Priority is introduced as an
+independent concept; meaning that any QoS class may have any valid priority.
+When a node is out of resources and pods needs to be preempted, we give
```

Comment:
We propose a new concept of priority which take precedence over QoS class. When it comes to preempting two pods with equal priority, the one with lower QoS tier may be preempted, but if the priorities are not equal, the one with lower priority will be preempted regardless of its QoS class.

> The concern is that if a guaranteed pod can be evicted because of a high priority besteffort pod, the guaranteed qos doesn't mean much.

@liggitt Even if we took QoS into account, a higher priority guaranteed pod could preempt a lower priority guaranteed pod. After the introduction of priority, guaranteed QoS only means that the pod is going to get its requested resources if it schedules. The QoS class by itself does not guarantee scheduling (this is true even today).

> Also, it would be unclear how many guaranteed pods should be preempted in order to free sufficient resources for a best effort pod that does not declare a resource request.

A besteffort pod does not preempt any pod to acquire resources. A besteffort pod is asking for zero resources semantically. As is the case today, a besteffort pod may be scheduled on a node with no resources and be starved. However, a besteffort pod may preempt other pods for reasons other than resource. For example, it may preempt another lower priority pod so that the affinity rules of the burstable pod can be satisfied.

@k82cn Please see my comment above that priority has precedence over QoS class.